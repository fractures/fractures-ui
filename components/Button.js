import { Component } from "react"
import cc from "classcat"

const defaultProps = {
	action: () => null,
	isActive: false,
	isRounded: false,
	type: "primary",
	value: ""
}

class Button extends Component {
	render() {
		const buttonClasses = cc({
			"fr-btn": true,
			"fr-btn-primary": this.props.type === "primary",
			"fr-btn-secondary": this.props.type === "secondary",
			"fr-btn--small": this.props.small,
			"fr-btn--disabled": this.props.disabled,
			"fr-btn--active": this.props.isActive,
			"fr-btn--radius": !this.props.isRounded,
			"radius-max": this.props.isRounded
		})

		return (
			<div className="flex flex-ycenter" onClick={ () => this.props.action() }>
				<input className={ buttonClasses } type="button" value={ this.props.value } />
				<style jsx global>{`
					.fr-btn {
						display: inline-block;
						padding: 0.25rem 1rem;

						border: 0;
						outline: none;

						font-size: 0.875rem;
						font-weight: 500;
						line-height: 2rem;
						cursor: pointer;

						transition: background-color 100ms ease-in-out, box-shadow 100ms ease-in-out;
					}

					.fr-btn-primary {
						background-color: var(--shade-4);
						border: 1px solid var(--shade-4);
						color: var(--shade-0);
					}

					.fr-btn-primary:not(.fr-btn--active):hover {
						background-color: var(--shade-6);
						border-color: var(--shade-6);
					}

					.fr-btn.fr-btn-primary.fr-btn--active,
					.fr-btn.fr-btn-primary:active {
						background-color: var(--shade-7);
						border-color: var(--shade-7);
						color: var(--shade-1);
					}

					.fr-btn-secondary {
						background-color: var(--shade-0);
						border: 1px solid var(--shade-4);
						color: var(--shade-4);
					}

					.fr-btn-secondary:not(.fr-btn--active):hover {
						background-color: var(--shade-1);
						border-color: var(--shade-3);
						color: var(--shade-5);
					}

					.fr-btn.fr-btn-secondary.fr-btn--active,
					.fr-btn.fr-btn-secondary:active {
						background-color: var(--shade-1);
						border-color: var(--shade-4);
						color: var(--shade-6);
					}

					.fr-btn--radius {
						border-radius: 0.25rem;
					}

					.fr-btn.fr-btn--small {
						padding: 0.125rem 0.5rem;

						font-size: 0.75rem;
						line-height: 1.5rem;
					}

					.fr-btn--disabled {
						opacity: 0.6;

						border-color: transparent;

						pointer-events: none;
					}

					.fr-btn:focus:not(.fr-btn--disabled):not(.fr-btn--loading) {
						box-shadow: 0 0 0 0.15rem var(--shade-2);
					}
				`}</style>
			</div>
		)
	}
}

Button.defaultProps = defaultProps

export default Button
