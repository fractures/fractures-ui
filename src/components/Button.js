import cc from "classcat"
import PropTypes from "prop-types"
import React from "react"

const propTypes = {
	action: PropTypes.func.isRequired,
	isActive: PropTypes.bool,
	isLoading: PropTypes.bool,
	isRounded: PropTypes.bool,
	type: PropTypes.oneOf(["primary", "secondary"]),
	value: PropTypes.string.isRequired
}

const defaultProps = {
	action: () => null,
	isActive: false,
	isLoading: false,
	isRounded: null,
	type: "primary",
	value: ""
}

const IconLoading = props => (
	<svg xmlns="http://www.w3.org/2000/svg" className="fr-btn__loading">
		<circle fill="none" strokeWidth={ 4 } stroke="var(--fr-300)" cx="12" cy="12" r="10" style={ { opacity: 0.25 } } />
		<circle
			fill="none"
			strokeWidth={ 4 }
			stroke={ props.color || "var(--fr-100)" }
			strokeDasharray="32"
			strokeLinecap="round"
			cx="12"
			cy="12"
			r="10"
		/>
	</svg>
)

const Button = props => {
	const buttonClasses = cc({
		"fr-btn": true,
		"fr-btn-primary": props.type === "primary",
		"fr-btn-secondary": props.type === "secondary",
		"fr-btn--small": props.small,
		"fr-btn--disabled": props.disabled,
		"fr-btn--loading": props.isLoading,
		"fr-btn--active": props.isActive,
		"fr-btn--radius": !props.isRounded,
		"radius-max": props.isRounded
	})

	return (
		<div className="inline-flex flex-ycenter relative" onClick={ e => props.action(e) }>
			<input className={ buttonClasses } type="button" value={ props.value } />
			{props.isLoading && <IconLoading color={ props.type === "secondary" && "var(--fr-500)" } />}
			<style jsx global>{`
				.fr-btn {
					display: inline-block;
					padding: 0.25rem 1rem;

					border: 0;
					outline: none;

					font-size: 0.875rem;
					font-weight: 500;
					line-height: 2rem;
					cursor: pointer;

					transition: background-color 100ms ease-in-out, box-shadow 100ms ease-in-out,
						padding-left 180ms ease-in-out;
				}

				/* Primary */
				.fr-btn-primary {
					background-color: var(--fr-500);
					color: var(--fr-ground);
				}

				.fr-btn-primary:not(.fr-btn--active):not(.fr-btn--loading):hover {
					background-color: var(--fr-700);
				}

				.fr-btn.fr-btn-primary.fr-btn--active,
				.fr-btn.fr-btn-primary:active,
				.fr-btn.fr-btn-primary:active:focus {
					background-color: var(--fr-900);
					color: var(--fr-100);
				}

				.fr-btn-primary.fr-btn--loading {
					color: var(--fr-300);
				}

				.fr-btn-primary.fr-btn--disabled {
					background-color: var(--fr-100);
					color: var(--fr-300);
				}

				/* Secondary */
				.fr-btn-secondary {
					background-color: var(--fr-ground);
					box-shadow: inset 0 0 0 1px var(--fr-300);
					color: var(--fr-500);
				}

				.fr-btn-secondary:not(.fr-btn--active):not(.fr-btn--loading):hover {
					box-shadow: inset 0 0 0 1px var(--fr-500);
					color: var(--fr-700);
				}

				.fr-btn.fr-btn-secondary.fr-btn--active,
				.fr-btn.fr-btn-secondary:active {
					background-color: var(--fr-100);
					box-shadow: inset 0 0 0 1px var(--fr-500);
					color: var(--fr-900);
				}

				.fr-btn-secondary.fr-btn--disabled {
					box-shadow: inset 0 0 0 1px var(--fr-100);
					color: var(--fr-300);
				}

				/* Default radius */
				.fr-btn--radius {
					border-radius: 0.25rem;
				}

				/* Small */
				.fr-btn.fr-btn--small {
					padding: 0.125rem 0.5rem;

					font-size: 0.75rem;
					line-height: 1.5rem;
				}

				/* Disabled */
				.fr-btn--disabled {
					pointer-events: none;
				}

				/* Loading */
				.fr-btn--loading {
					padding-left: 2.5rem;

					pointer-events: none;
				}

				.fr-btn--small.fr-btn--loading {
					padding-left: 2rem;
				}

				.fr-btn--small + .fr-btn__loading {
					left: 0.125rem;
				}

				.fr-btn__loading {
					position: absolute;
					left: 0.5rem;
					top: 0.5rem;
					z-index: 1;

					height: 1.5rem;
					width: 1.5rem;

					animation: fr-button-show 180ms ease-in-out;
				}

				.fr-btn__loading circle {
					animation: fr-button-dash 480ms linear forwards infinite;
				}

				@keyframes fr-button-show {
					0% {
						opacity: 0;
					}
					100% {
						opacity: 1;
					}
				}

				@keyframes fr-button-dash {
					from {
						stroke-dashoffset: 0;
					}
					to {
						stroke-dashoffset: 64;
					}
				}

				/* Focus */
				.fr-btn:focus {
					z-index: 1;
				}

				.fr-btn.fr-btn--small:focus {
					box-shadow: 0 0 0 0.125rem var(--fr-focus);
				}

				.fr-btn:focus:not(.fr-btn--disabled):not(.fr-btn--loading):not(.fr-btn--small) {
					box-shadow: 0 0 0 0.1875rem var(--fr-focus);
				}

				.fr-btn.fr-btn-secondary:focus:not(.fr-btn--disabled):not(.fr-btn--loading) {
					box-shadow: inset 0 0 0 1px var(--fr-500), 0 0 0 0.1875rem var(--fr-focus);
				}
			`}</style>
		</div>
	)
}

Button.propTypes = propTypes
Button.defaultProps = defaultProps

export default Button
