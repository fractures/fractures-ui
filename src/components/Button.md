### Primary buttons

```js noeditor
<Fractures theme="blue" />
```

```js
<Button value="Primary button" type="primary" />
```

### Secondary button

```js
<Button value="Secondary button" type="secondary" />
```
