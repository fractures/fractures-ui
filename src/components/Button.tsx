import ButtonPrimary from './ButtonPrimary'
import ButtonSecondary from './ButtonSecondary'
import classNames from '@sindresorhus/class-names'
import LoadingIcon from './LoadingIcon'
import React from 'react'

const Button = props => {
	const buttonClasses = classNames({
		isActive: props.isActive,
		isDisabled: props.isDisabled,
		isLoading: props.isLoading,
		isRound: props.isRounded,
		isSmall: props.isSmall
	})

	const LoadingColor = props.type === 'secondary' ? 'var(--color-gray-300)' : 'var(--color-gray-100)'

	return (
		<div style={ { position: 'relative', display: 'inline-flex', alignItems: 'center' } } onClick={ e => props.onClick(e) }>
			{props.type === 'primary' && <ButtonPrimary { ...props } className={ buttonClasses } />}
			{props.type === 'secondary' && <ButtonSecondary { ...props } className={ buttonClasses } />}
			{props.isLoading && <LoadingIcon color={ LoadingColor } size={ props.isSmall ? 10 : 20 } />}
		</div>
	)
}

export default Button
