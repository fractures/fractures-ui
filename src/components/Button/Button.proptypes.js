import PropTypes from "prop-types"

const propTypes = {
	action: PropTypes.func.isRequired,
	className: PropTypes.string,
	isActive: PropTypes.bool,
	isDisabled: PropTypes.bool,
	isLoading: PropTypes.bool,
	isRounded: PropTypes.bool,
	isSmall: PropTypes.bool,
	type: PropTypes.oneOf(["primary", "secondary"]),
	value: PropTypes.string.isRequired
}

const defaultProps = {
	action: () => null,
	isActive: false,
	isDisabled: false,
	isLoading: false,
	isRounded: null,
	type: "primary",
	value: ""
}

export { propTypes, defaultProps }
