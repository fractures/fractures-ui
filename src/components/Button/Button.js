import { propTypes, defaultProps } from "./Button.proptypes"
import ButtonPrimary from "./components/ButtonPrimary"
import ButtonSecondary from "./components/ButtonSecondary"
import classNames from "@sindresorhus/class-names"
import LoadingIcon from "./components/LoadingIcon"
import React from "react"

const Button = props => {
	const buttonClasses = classNames({
		isActive: props.isActive,
		isDisabled: props.isDisabled,
		isLoading: props.isLoading,
		isRound: props.isRounded,
		isSmall: props.isSmall
	})

	const LoadingColor = props.type === "secondary" ? "var(--color-gray-300)" : "var(--color-gray-100)"

	return (
		<div style={ { position: "relative", display: "inline-flex", alignItems: "center" } } onClick={ e => props.onClick(e) }>
			{props.type === "primary" && <ButtonPrimary { ...props } className={ buttonClasses } />}
			{props.type === "secondary" && <ButtonSecondary { ...props } className={ buttonClasses } />}
			{props.isLoading && <LoadingIcon color={ LoadingColor } size={ props.isSmall ? 10 : 20 } />}
		</div>
	)
}

Button.propTypes = propTypes
Button.defaultProps = defaultProps

export default Button
