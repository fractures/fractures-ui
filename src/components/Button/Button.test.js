import Button from "./Button.js"

const mockFn = jest.fn()

describe("Button", () => {
	const componentRender = <Button value="Primary" action={ mockFn(true) } />
	const component = renderer.create(componentRender)
	const tree = component.toJSON()
	const app = mount(componentRender)

	it("Renders correctly", () => {
		expect(tree).toMatchSnapshot()
	})

	it("Click on button actually calls passed function", () => {
		const button = app.find("input")

		button.simulate("click")
		expect(mockFn).toBeCalledWith(true)
	})
})

describe("Button Secondary", () => {
	const componentRender = <Button value="Secondary" type="secondary" action={ mockFn(false) } />
	const component = renderer.create(componentRender)
	const tree = component.toJSON()
	const app = mount(componentRender)

	it("Renders correctly", () => {
		expect(tree).toMatchSnapshot()
	})

	xit("Secondary button gets secondary class", () => {
		expect(app.find(".hl-btn").hasClass("hl-btn-secondary")).toEqual(true)
	})
})
