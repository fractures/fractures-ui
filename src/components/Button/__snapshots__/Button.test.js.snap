// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Renders correctly 1`] = `
<div
  onClick={[Function]}
  style={
    Object {
      "alignItems": "center",
      "display": "inline-flex",
      "position": "relative",
    }
  }
>
  <input
    className="sc-bwzfXH kYTxEm"
    role="button"
    type="button"
    value="Primary"
  />
</div>
`;

exports[`Button Secondary Renders correctly 1`] = `
<div
  onClick={[Function]}
  style={
    Object {
      "alignItems": "center",
      "display": "inline-flex",
      "position": "relative",
    }
  }
>
  <input
    className="sc-htpNat fYIdIk"
    role="button"
    type="button"
    value="Secondary"
  />
</div>
`;
